### PR TITLE
just buildProfile

### PR DESCRIPTION
**Pull request by kgrosman**
buildProfile script to build a profile form a base and using specified sections of a CUJ. This allows us to easily build and test complex profiles including actions from CUJS that don't start with app startup.

### CODE REVIEW

File: CUJExecutor/buildProfile [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+echo "Resetting the profile...."
+adb shell cmd package compile --reset com.google.android.apps.maps
+
+echo "Setting the base profile..."
+#If we're given a file name, set that to be the profile we use and exit
+if [ -f "$COMPILATION_FLAG_SEQ" ] 
+then 
+	
+	FILE_EXTENSION="${COMPILATION_FLAG_SEQ##*.}"
```
Comment (diablonhn) [inline CUJExecutor/buildProfile]: Not too familar with advanced uses of bash.  How does this extract out the file extension?

File: CUJExecutor/buildProfile [hunk 2]
```diff
@@ -0,0 +1,117 @@
+#!/usr/bin/env bash
+	
+
+INPUT_FILE=${1?"ERROR: need to provide an input file and a compilation flag"}
+COMPILATION_FLAG_SEQ=${2?"ERROR: need to provide an input file and a compilation flag"}
+COMMAND_SUMMARY="/tmp/build_command_summary.txt"
+CLOUD_PROFILE="utilityProfiles/cloudProfile.prof" #location of the saved cloud profile
+EMPTY_PROFILE="utilityProfiles/empty.prof" #location of an empty profile
+STARTUP_PROFILE="utilityProfiles/startup.prof" #location of an empty profile
```
Comment (diablonhn) [inline CUJExecutor/buildProfile]: Update comment for STARTUP_PROFILE:

    #location of an empty profile

File: CUJExecutor/buildProfile [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+	cp $EMPTY_PROFILE $STASHING_CREATED_PROFILE
+	echo -n "Base: Empty, Actions:" > $STASHING_COMPILED_ACTIONS
+elif [ "$BASE_PROFILE" = "u" ]
+then
+	cp $CLOUD_PROFILE $STASHING_CREATED_PROFILE	
+	echo -n "Base: Cloud, Actions:" > $STASHING_COMPILED_ACTIONS
+elif [ "$BASE_PROFILE" = "s" ]
+then
+	cp $STARTUP_PROFILE $STASHING_CREATED_PROFILE	
+	echo -n "Base: Startup, Actions:" > $STASHING_COMPILED_ACTIONS
```
Comment (diablonhn) [inline CUJExecutor/buildProfile]: Will the base profile always be the startup profile?  Or do we want to be able to specify an arbitrary profile?

File: CUJExecutor/buildProfile [hunk 4]
```diff
@@ -0,0 +1,117 @@
+#!/usr/bin/env bash
+	
+
+INPUT_FILE=${1?"ERROR: need to provide an input file and a compilation flag"}
+COMPILATION_FLAG_SEQ=${2?"ERROR: need to provide an input file and a compilation flag"}
```
Comment (diablonhn) [inline CUJExecutor/buildProfile]: Not sure what `COMPILATION_FLAG_SEQ stands for. Maybe a better name?

It looks like a list of flags.  Add examples to the error message or as a comment?

File: CUJExecutor/buildProfile [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+	then	
+		echo "Builder actions flag: $COMPILATION_FLAG is invalid"
+		exit 1
+	fi
+
+	if [ "$COMPILATION_FLAG" = "m" ] 
+	then 
+		
+		echo "Executing preparatory actions.."
+		./runCujNTimes $INPUT_FILE p 1
```
Comment (diablonhn) [inline CUJExecutor/buildProfile:88]: Should add a comment for what `p` and `1` are.

File: CUJExecutor/buildProfile [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+		echo "Waiting for prepatory actions to be added to the profile..."
+		sleep 60
+	fi
+	
+
+	echo "Pushing our working profile"
+	adb push $STASHING_CREATED_PROFILE $CUR_PROFILE
+
+	echo "Restoring profile permissions"
+	adb shell chmod a+rw $CUR_PROFILE
```
Comment (diablonhn) [inline CUJExecutor/buildProfile]: I believe adb push messes up the file owner and permissions. May want to push to a temporary location and then do an adb shell cp to keep the existing permissions.

File: CUJExecutor/buildProfile [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+	fi
+
+	echo "Pulling the current profile"
+	adb pull $CUR_PROFILE $STASHING_CREATED_PROFILE
+
+	echo "Getting profile size..."
+	adb shell stat -c%s $CUR_PROFILE
+
+	echo "Finishing the CUJ..."
+	./runCujNTimes $INPUT_FILE ${COMPILATION_FLAG}r 1
```
Comment (diablonhn) [inline CUJExecutor/buildProfile:121]: Should add a comment for `${COMPILATION_FLAG}r`.

File: CUJExecutor/buildProfile [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+
+
+	echo "Force stopping Maps..." #Want to make sure that app is stopped when we mess with the profiles at start of the loop in callee
```
Comment (diablonhn) [inline CUJExecutor/buildProfile]: Generally easier to read if comment is right above the code, as opposed to the right.

File: CUJExecutor/manipulateProfilesAndTest [hunk 9]
```diff
@@ -19,36 +19,34 @@
...
 # - (saved_profile.prof) the profile used to compile the app (if applicable). 
 # - (median_clip.mp4) a recording of the median iteration (if requested).
 #
-# @param INPUT_FILE (Formatted with the following on consecutive lines: PRE-CUJ, POST_CUJ, NUM_ITERS, and optionally RECORD (a line with an "r" to indicate that you'd like to record). See CUJS/ for an example file.
-# @param COMPILATION_FLAG (e - empty profile, p - compile actions in Pre-CUJ, c - compile entire CUJ, a - compile all code).
+# @param INPUT_FILE (Formatted with the following on consecutive lines: PRE-CUJ, MEASURED_CUJ, POST_CUJ, NUM_ITERS, and optionally RECORD (a line with an "r" to indicate that you'd like to record). See CUJS/ for an example file.
+# @param COMPILATION_FLAG (e - empty profile, m - compile measured CUJ, u - compile with a cloud profile, f- compile first action, p - compile actions in Pre-CUJ, c - compile pre and measured CUJS, a - compile all code).
+#  	 Optionally, if f, p, c, or m are followed by a number, n, (e.g. m12), that portion of the CUJ will be run through n times during the creation of the profile. If a, e or u are followed by a number, that number will be ignored.
```
Comment (diablonhn) [inline CUJExecutor/manipulateProfilesAndTest]: No action requested.  Pretty neat, flexible, and powerful (e.g. m12).  But very succinct.

**Review by diablonhn — CHANGES_REQUESTED** [review]

File: CUJExecutor/buildProfile (hunk 3 above)
Comment (kgrosman) [reply CUJExecutor/buildProfile]: Base profile is determined based on the first flag in the sequence

File: CUJExecutor/buildProfile (hunk 6 above)
Comment (kgrosman) [reply CUJExecutor/buildProfile]: messes up the permission on the device? Would chmod a+rw fix that?

File: CUJExecutor/buildProfile (hunk 1 above)
Comment (kgrosman) [reply CUJExecutor/buildProfile]: Basically pattern matches the string to "*." and then returns the deleted substring, which is everything after the period.

File: CUJExecutor/buildProfile (hunk 3 above)
Comment (diablonhn) [reply CUJExecutor/buildProfile]: Right.  Would someone provide an arbitrary profile by specifying it as the startup or cloud profile?

File: CUJExecutor/buildProfile (hunk 6 above)
Comment (diablonhn) [reply CUJExecutor/buildProfile]: `chmod` would give access, but `adb push` changes the file owner and group to be root.  It's better to reuse the existing file permissions, which `adb shell cp` does.

**Review by diablonhn — APPROVED** [review]

File: CUJExecutor/buildProfile (hunk 3 above)
Comment (kgrosman) [reply CUJExecutor/buildProfile]: That would work. At the moment they can also just provide a file location and that profile will be used, but they can't build on top of it. If we see a use for it, I could change the format slightly such that you can use flags to indicate that you want to build on top of an arbitrary profile.

**Pull request merged**